infrastructure/storage/src/key_val_store/key_val_store.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright 2019. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::key_val_store::KeyValStoreError;
+use lmdb_zero::{traits::AsLmdbBytes, Stat};
+
+/// General CRUD behaviour of Key-value store implementations.
+pub trait KeyValStore {
+    /// Inserts a key-value pair into the key-value database.
+    fn insert_pair<K, V>(&self, key: &K, value: &V) -> Result<(), KeyValStoreError>
+    where
+        K: AsLmdbBytes + ?Sized,
+        V: serde::Serialize;
+
+    /// Get the value corresponding to the provided key from the key-value database.
+    fn get_value<K, V>(&self, key: &K) -> Result<Option<V>, KeyValStoreError>
+    where
+        K: AsLmdbBytes + ?Sized,
+        for<'t> V: serde::de::DeserializeOwned;
+
+    /// Return statistics about the key-value database, See [Stat](lmdb_zero/struct.Stat.html) for more details.
+    fn get_stats(&self) -> Result<Stat, KeyValStoreError>;
+
+    /// Log some pretty printed stats about the key-value database, See [Stat](lmdb_zero/struct.Stat.html) for more
+    /// details.
+    fn log_info(&self);
```

Comment:
I feel that the KeyValueStore should only have methods common to a key value store (basically CRUD methods) and leave out the LMDB specific functionality. 

Having log_info and get_stats as part of the trait has the disadvantage of forcing the implementer to "make up" stats if the backend doesn't support it. Would a user of this API (i.e having the KeyValueStore trait bound) generally have a use for calling "log_info"? (maybe get_stats, but in that case you probably want to only use the concrete LMDB type, or put an additional trait bound for a different trait such as `DatabaseStats` and have only LMDB implement that) - traits should have a single concern. 

For a hashmap implementation, "getting LMDB stats" doesn't make sense.

Reply:
Whoops - ^ should have gone on #490 ;)

Reply:
I agree. This trait was written to mimic the LMDBDatabase, but I see those two methods are not used so I will removed them.